src/compile/data/data.ts

```diff
@@ -79,61 +83,151 @@ export interface SummaryComponent {
 // TODO: split this file into multiple files and remove this linter flag
 /* tslint:disable:no-use-before-declare */
 
-export function parseUnitData(model: UnitModel): DataComponent {
+/**
+ * Parses a model with data.
+ */
+function parseData(model: Model): DataComponent {
   return {
+    source: source.parseUnit(model),
     formatParse: formatParse.parseUnit(model),
-    nullFilter: nullFilter.parseUnit(model),
+    calculate: calculate.parseUnit(model),
     filter: filter.parseUnit(model),
-    nonPositiveFilter: nonPositiveFilter.parseUnit(model),
-
-    source: source.parseUnit(model),
     bin: bin.parseUnit(model),
-    calculate: formula.parseUnit(model),
     timeUnit: timeUnit.parseUnit(model),
-    timeUnitDomain: timeUnitDomain.parseUnit(model),
-    summary: summary.parseUnit(model),
+    aggregate: aggregate.parseUnit(model),
+    colorRank: colorRank.parseUnit(model),
     stackScale: stackScale.parseUnit(model),
-    colorRank: colorRank.parseUnit(model)
+    nonPositiveFilter: nonPositiveFilter.parseUnit(model),
+    nullFilter: nullFilter.parseUnit(model),
+    timeUnitDomain: timeUnitDomain.parseUnit(model)
   };
 }
 
+/**
+ * Create components for data source and transforms/ aggregates.
+ */
+export function parseUnitData(model: UnitModel): DataComponent {
+  return parseData(model);
+}
+
+/**
+ * The data component for a facet is the data component of its child.
+ * We only need to add the facet fields as dimensions to the summary component.
+ */
 export function parseFacetData(model: FacetModel): DataComponent {
-  return {
-    formatParse: formatParse.parseFacet(model),
-    nullFilter: nullFilter.parseFacet(model),
-    filter: filter.parseFacet(model),
-    nonPositiveFilter: nonPositiveFilter.parseFacet(model),
-
-    source: source.parseFacet(model),
-    bin: bin.parseFacet(model),
-    calculate: formula.parseFacet(model),
-    timeUnit: timeUnit.parseFacet(model),
-    timeUnitDomain: timeUnitDomain.parseFacet(model),
-    summary: summary.parseFacet(model),
-    stackScale: stackScale.parseFacet(model),
-    colorRank: colorRank.parseFacet(model)
-  };
+  // TODO: treat data in the facet and in the child differently
+
+  const child = model.child();
+  const dataComponent = child.component.data;
+  delete child.component.data;
+
+  [SOURCE, RAW, RANK, SCALE, STACKED_SCALE].forEach((data) => {
+    model.renameData(child.dataName(data), model.dataName(data));
+  });
+
+  if (dataComponent.aggregate) {
+    aggregate.parseFacet(model, dataComponent.aggregate);
+  }
+  if (dataComponent.stackScale) {
+    stackScale.parseFacet(model, dataComponent.stackScale);
+  }
+
+  return dataComponent;
 }
 
+function mergeChildren(model: Model, dataComponent: DataComponent, children: UnitModel[]): DataComponent {
```

Comment:
Right now we employ all or nothing model.  I wonder if we can adjust a little bit by removing problematic `childDataComponent` from `childDataComponents`.  This way the code shouldn't be that much more complicated, but we can still reuse the logic for optimizing in many cases.  

(Maybe future work.)